### PR TITLE
lite2: replace primary when providing invalid header

### PR DESCRIPTION
**Pull request by cmwaters**
Closes: #4420

## Description

Created a new error `ErrInvalidHeader`which can be formed during the verification process `verifier.go` and will result in the replacement of the primary provider with a witness by executing: `replacePrimaryProvider()`

______

For contributor use:

- [ ] Wrote tests
- [ ] Updated CHANGELOG_PENDING.md
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: lite2/errors.go [hunk 1]
```diff
@@ -28,3 +28,11 @@ type ErrNewValSetCantBeTrusted struct {
 func (e ErrNewValSetCantBeTrusted) Error() string {
 	return fmt.Sprintf("cant trust new val set: %v", e.Reason)
 }
+
+type ErrInvalidHeader struct {
+	Reason string
```
Comment (melekes) [inline lite2/errors.go]:
```suggestion
	Reason error
```

File: lite2/client.go [hunk 2]
```diff
@@ -711,7 +711,19 @@ func (c *Client) sequence(
 		err = VerifyAdjacent(c.chainID, trustedHeader, interimHeader, interimVals,
 			c.trustingPeriod, now)
 		if err != nil {
-			return errors.Wrapf(err, "failed to verify the header #%d", height)
+			switch err.(type) {
+			case ErrInvalidHeader:
+				c.logger.Debug("replacing primary because: ", "err", err.Error())
```
Comment (melekes) [inline lite2/client.go]:
```suggestion
				c.logger.Error("primary sent us invalid header", "err", err, "primary", primary)
```

File: lite2/client.go [hunk 3]
```diff
@@ -711,7 +711,19 @@ func (c *Client) sequence(
...
+				err = c.replacePrimaryProvider()
```
Comment (melekes) [inline lite2/client.go]:
```suggestion
				err1 := c.replacePrimaryProvider()
```

File: lite2/client.go [hunk 4]
```diff
@@ -711,7 +711,19 @@ func (c *Client) sequence(
...
+				if err != nil {
+					return errors.Wrapf(err, "tried to verify adjacent header")
```
Comment (melekes) [inline lite2/client.go]:
```suggestion
					err = errors.Wrapf(err, "failed to verify the header #%d", height)
					return errors.Wrapf(err1, "can't replace primary (original error: %v)", err)
```

File: lite2/client.go [hunk 5]
```diff
@@ -830,19 +851,36 @@ func (c *Client) backwards(
 		}
 
 		if err := interimHeader.ValidateBasic(c.chainID); err != nil {
```
Comment (melekes) [inline lite2/client.go]: the following 3 if conditions need to be extracted into a function. then we won't have to repeat replacePrimaryProvider 3 times

Comment (cmwaters) [reply lite2/client.go]: I'm going to put it in `verifier.go` unless you disagree

File: lite2/client.go [hunk 6]
```diff
@@ -711,7 +711,20 @@ func (c *Client) sequence(
 		err = VerifyAdjacent(c.chainID, trustedHeader, interimHeader, interimVals,
 			c.trustingPeriod, now)
 		if err != nil {
-			return errors.Wrapf(err, "failed to verify the header #%d", height)
+			switch err.(type) {
+			case ErrInvalidHeader:
+				c.logger.Error("replacing primary because: ", "err", err.Error())
```
Comment (melekes) [inline lite2/client.go]:
```suggestion
				c.logger.Error("primary sent us invalid header -> replacing", "err", err)
```

File: lite2/client.go [hunk 7]
```diff
@@ -764,6 +777,16 @@ func (c *Client) bisection(
 				return err
 			}
 
+		case ErrInvalidHeader:
+			c.logger.Error("replacing primary because: ", "err", err.Error())
```
Comment (melekes) [inline lite2/client.go]:
```suggestion
			c.logger.Error("primary sent us invalid header -> replacing", "err", err)
```

File: lite2/verifier.go [hunk 8]
```diff
@@ -118,7 +118,27 @@ func VerifyAdjacent(
...
 	if err := untrustedVals.VerifyCommit(chainID, untrustedHeader.Commit.BlockID, untrustedHeader.Height,
 		untrustedHeader.Commit); err != nil {
-		return err
+		return ErrInvalidHeader{err}
+	}
+
+	return nil
+}
+
+func VerifyBackwards(chainID string, untrustedHeader, trustedHeader *types.SignedHeader) error {
```
Comment (melekes) [inline lite2/verifier.go]: godoc please

**Comment by melekes** [comment]
Should we write tests for this?

**Comment by cmwaters** [comment]
> Should we write tests for this?

All the verification tests are already covered - the only thing we would need to check is that the providers are replaced and that the verification process resumes as normal

**Comment by cmwaters** [comment]
I wrote some more helper functions in the benchmarking PR that we could use if we want to construct more elaborate providers

**Comment by melekes** [comment]
> I wrote some more helper functions in the benchmarking PR that we could use if we want to construct more elaborate providers

works for me

**Pull request merged**